// @lagoshny/ngx-hateoas-client/esm2020/lib/config/lib-config.mjs
import { DEFAULT_ROUTE_NAME } from './hateoas-configuration.interface';
import { CacheMode } from '../model/declarations';
/**
 * Contains all configuration lib params.
 */
// tslint:disable:no-string-literal
export class LibConfig {
    static setConfig(hateoasConfiguration) {
        LibConfig.config = LibConfig.mergeConfigs(hateoasConfiguration);
    }
    static getConfig() {
        return LibConfig.config;
    }
    static mergeConfigs(config) {
        return {
            ...LibConfig.DEFAULT_CONFIG,
            ...config,
            cache: {
                ...LibConfig.DEFAULT_CONFIG.cache,
                ...config.cache
            },
        };
    }
}
LibConfig.DEFAULT_CONFIG = {
    http: {
        [DEFAULT_ROUTE_NAME]: {
            rootUrl: 'http://localhost:8080/api/v1'
        },
    },
    logs: {
        verboseLogs: false
    },
    cache: {
        enabled: true,
        mode: CacheMode.ALWAYS,
        lifeTime: 5 * 60 * 1000
    },
    useTypes: {
        resources: []
    },
    pagination: {
        defaultPage: {
            size: 20,
            page: 0
        }
    },
    halFormat: {
        collections: {
            embeddedOptional: false
        }
    },
    isProduction: false
};
LibConfig.config = LibConfig.DEFAULT_CONFIG;
