// @lagoshny/ngx-hateoas-client/esm2020/lib/ngx-hateoas-client.module.mjs
import { NgModule } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./config/ngx-hateoas-client-configuration.service";
export { NgxHateoasClientConfigurationService } from './config/ngx-hateoas-client-configuration.service';
export { Resource } from './model/resource/resource';
export { EmbeddedResource } from './model/resource/embedded-resource';
export { ResourceCollection } from './model/resource/resource-collection';
export { PagedResourceCollection } from './model/resource/paged-resource-collection';
export { Include, HttpMethod, CacheMode } from './model/declarations';
export { HateoasResourceOperation } from './service/external/hateoas-resource-operation';
export { HateoasResourceService } from './service/external/hateoas-resource.service';
export { HateoasResource, HateoasEmbeddedResource, HateoasProjection, ProjectionRel } from './model/decorators';
export class NgxHateoasClientModule {
    constructor(config) {
    }
    static forRoot() {
        return {
            ngModule: NgxHateoasClientModule
        };
    }
}
NgxHateoasClientModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NgxHateoasClientModule, deps: [{ token: i1.NgxHateoasClientConfigurationService }], target: i0.ɵɵFactoryTarget.NgModule });
NgxHateoasClientModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NgxHateoasClientModule });
NgxHateoasClientModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NgxHateoasClientModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: NgxHateoasClientModule, decorators: [{
            type: NgModule
        }], ctorParameters: function () { return [{ type: i1.NgxHateoasClientConfigurationService }]; } });
